### DISCUSSION TITLE
Tiny Web Application Framework

### DISCUSSION THREAD

**Discussion by DavesCodeMusings** [micropython/micropython#9571] [Show and tell]
I'm putting together a web framework similar in spirit to Python Flask. You can find the code on [here on GitHub](https://github.com/DavesCodeMusings/thimble) Here's an example that may look familiar if you've used Flask:

```
from thimble import Thimble

app = Thimble() 

@app.route('/')
def hello():
    return 'Hello!'

app.run()
```

Point your browser to your micro-controller's IP address and see the magic.

Obviously, it's going to be very light on functionality to fit on a micro-controller. But, I think it's a nice step up from the typical web server example code.

My plan is to get it to the point where it's suitable for building a REST API component for micro-controller projects. (Like controlling one of several GPIOs by sending a POST or PUT to a particular URL path.)

Currently, it can do GET requests like this:

```
from thimble import Thimble

app = Thimble('text/plain')

@app.route('/')
def hello():
    return 'Hello!'

@app.route('/world')
def hello_cleveland():
    return 'Hello World!'

@app.route('/cleveland')
def hello_cleveland():
    return 'Hello Cleveland!'

app.run()  # Listen on 0.0.0.0:80 by default.
```

So if you wanted to do something like reading temperature, humidity and barometric pressure from a BME280 sensor using HTTP URLs in the format of /temperature/celsius or /pressure/millibars, it should be suitable for that.

Still need some work to handle POST, PUT, etc. But, if you like Flask, and you like MicroPython too, this might be the project for you. Try it out and see.

**Comment by DoubleVee73** [comment]
Thanks! This looks very promising. I am looking to develop a product that is microcontroller based and presents a REST API using HTTPS. I will follow this with interest thank you

**Reply by DavesCodeMusings** [reply]
I don't know that I'll go to the extent of adding TLS (HTTPS) support, but thanks for following.

**Comment by mattytrentini** [comment]
Are you aware of [microdot](https://github.com/miguelgrinberg/microdot)? Not to discourage you, but it's also inspired by Flask and has a near-identical interface to what you're proposing. It's also quite mature and, despite it's small size, has some rather powerful features like async support and templates.

**Reply by DavesCodeMusings** [reply]
I had not heard of microdot until now. It looks a lot like what I had envisioned. I'll probably keep going with my project though, because it's been a great learning experience.

**Reply by dpgeorge** [reply]
I agree that it's a great learning experience to write a web framework.  There are many aspects to learn from it.  I would encourage you to look at adding `uasyncio` support :smile:

**Reply by DavesCodeMusings** [reply]
uasyncio would definitely improve things. I'll need to wrap my head around it first as I've only done async in JavaScript thus far.

**Comment by DavesCodeMusings** [comment]
I've got it set up to run sync or asych and made some other improvements. I think it's pretty usable for slapping an API on top of a project. For example, here's the code for a nodemcu (8266) and DHT22 sensor with an API provided by the web framework.

```
# Example program using Thimble to create an API.
from machine import Pin
from dht import DHT22
from thimble import Thimble
import uasyncio

gpio_2 = Pin(2, Pin.OUT)  # Built-in LED is used as a nightlight.
gpio_2.value(1)  # Turn off the active-low device for now.

dht22 = DHT22(Pin(14))
temp_c = 0
temp_f = 32
humidity = 0

# Read values from DHT sensor periodically.
async def read_dht():
    global temp_c
    global temp_f
    global humidity

    while (True):
        dht22.measure()
        temp_c = dht22.temperature()
        temp_f = 1.8 * temp_c + 32
        humidity = dht22.humidity()
        print(f'Temp: {temp_c}C, {temp_f}F  Humidity: {humidity}%')
        await uasyncio.sleep(15)


# Instantiate the Thimble API object.
api = Thimble()

# Add routes for reading the DHT-provided values using the default HTTP method of GET.
@api.route('/temperature/celsius')
def get_temp_c(req):
    return round(temp_c)

@api.route('/temperature/fahrenheit')
def get_temp_f(req):
    return round(temp_f)

@api.route('/humidity')
def get_humidity(req):
    return round(humidity)

# Various HTTP methods allow control of the built-in LED.
@api.route('/nightlight', methods=['GET'])
def get_nightlight(req):
    return 'on' if gpio_2.value() == 0 else 'off'

@api.route('/nightlight', methods=['PUT'])
def set_nightlight(req):
    if (req['body'] == 'on'):
        gpio_2.value(0)  # Built-in LED is active-low, so 0 is on
    elif (req['body'] == 'off'):
        gpio_2.value(1)
    return 'on' if gpio_2.value() == 0 else 'off'

@api.route('/nightlight', methods=['POST'])
def set_nightlight_on(req):
    gpio_2.value(0)
    return 'on' if gpio_2.value() == 0 else 'off'

@api.route('/nightlight', methods=['DELETE'])
def set_nightlight_off(req):
    gpio_2.value(1)
    return 'on' if gpio_2.value() == 0 else 'off'


# Run the DHT reader and Thimble as asynchronous tasks.
loop = uasyncio.get_event_loop()
api.run_async(loop=loop)
uasyncio.create_task(read_dht())
loop.run_forever()
```

This is my first "big" MicroPython project, so any feedback is appreciated.

**Comment by andypiper** [comment]
Looks nice! Sent a [PR for Awesome MicroPython](https://github.com/mcauser/awesome-micropython/pull/73) 👍🏻

**Reply by DavesCodeMusings** [reply]
Thanks!
